libraries/AC_Avoidance/AP_OADatabase.cpp

```diff
@@ -0,0 +1,291 @@
+        return;
+    }
+
+    _queue.size = _queue_size_param;
+    _queue.items = new ObjectBuffer<OA_DbItem>(_queue.size);
+    if (_queue.items == nullptr) {
+        // We're unable to allocate memory with the smallest allowed size. hard failure, do not retry.
+        gcs().send_text(MAV_SEVERITY_INFO, "DB failed to allocate queue of size %d", _queue.size);
+        _unable_to_allocate_memory = true;
+    }
+}
+
+void AP_OADatabase::init_database()
+{
+    if (_database.items != nullptr || _database_size_param < _database.size_min) {
+        return;
+    }
+
+    _database.size = _database_size_param;
+    _database.items = (OA_DbItem *)calloc(_database.size, sizeof(OA_DbItem));
+    if (_database.items == nullptr) {
+        // We're unable to allocate memory with the smallest allowed size. hard failure, do not retry.
+        gcs().send_text(MAV_SEVERITY_INFO, "DB failed to allocate database of size %d - not retrying", _database.size);
+        _unable_to_allocate_memory = true;
+    }
+}
+
+void AP_OADatabase::optimize_db_filter()
+{
+    // TODO: check database size and if we're getting full
+    // we should grow the database size and/or increase
+    // _database_filter_m so less objects go into it and let
+    // the existing ones timeout naturally
+
+    if (_database.count > (_database.size * 0.90f)) {
+        // we're almost full, lets increase the filter size by requiring more
+        // spacing between points so less things get put into the database
+        _database.filter_m = MIN(_database.filter_m*_database.filter_grow_rate, _database.filter_max_m);
+
+    } else if (_database.count < (_database.size * 0.85f)) {
+        // we have some room, lets loosen the filter requirement (smaller object points) to allow more samples
+        _database.filter_m = MAX(_database.filter_m*_database.filter_shrink_rate,_database.filter_min_m);
+    }
+}
+
+void AP_OADatabase::process_queue()
+{
+    // processing queue by moving those entries into the database
+    // Using a for with fixed size is better than while(!empty) because the
+    // while could get us stuck here longer than expected if we're getting
+    // a lot of values pushing into it while we're trying to empty it. With
+    // the for we know we will exit at an expected time
+    const uint32_t queue_available = _queue.items->available();
+    for (uint32_t queue_index=0; queue_index<queue_available; queue_index++) {
+        OA_DbItem item;
+
+        bool pop_success;
+        {
+            WITH_SEMAPHORE(_queue.sem);
+            pop_success = _queue.items->pop(item);
+        }
+        if (!pop_success) {
+            break;
+        }
+
+        // compare item to all items in database. If found a similar item, update the existing, else add it as a new one
+        for (uint32_t i=0; i<_database.count; i++) {
+            if (is_item_in_database(i, item)) {
+                database_item_refresh(i, item.timestamp);
+                return;
```

Comment:
I don't think `return` is correct here. This makes you stop processing the outer loop, which means if there is a match on the first item in the queue, nothing else is pulled out, which introduces a large delay of ~100ms before we pull the next item out of the queue. If you are near a wall and using the MAVLink message this could easily lead to not updating your queue correctly.

Reply:
good catch, but it's still getting served with the 1ms delay that you didn't notice and thought was a busy loop.